### ISSUE TITLE
How to MultiTask Training (ComputeGraph) with different size of Task data set? 

### ISSUE THREAD

**Issue by yylonly** [deeplearning4j/deeplearning4j#3559]
=====================================================================

Network is 

--------->Task A
X - f ---- >Task B
---------> Task C

Dataset is here:

Dataset of TaskA : 500 numbers 
Dataset of TaskB : 100 numbers
Dataset of TaskC : 400 numbers

X in each task is different.

====================================================================

DL4J can build the multi-task network by ComputeGraph. the training can only happen when each task has the same size and same input x. But it can not be trained for the situation above.

Tensorflow provides low-level API can 
Session.run(inputA, outputA, lostA) 
Session.run(inputB, outputB, lostB)
Session.run(inputC, outputC, lostC)
on the same network. 

However, DL4J fit() function can not do it. 

For now, I use transfer learning API, training A first, then remove TaskA classification layers, save parameters of classification layer of TaskA, then add B classification layers, then training again................Then to C. 

That is a compromise solution. However, when the number of task is great than e.g 2000, that we can not handle it only in Transfer learning API...

I really wanna do multi-task in DL4J. if has any suggestion, I can contribute D4LJ. I really love it..

Thanks

**Comment by AlexDBlack** [comment]
Alright, so this is a case of multi-task learning.

Some options available in DL4J currently are:
1. Transfer learning API to create a new network, as you mentioned
2. For binary classification, or regression multi-task problems, use a single output layer (all outputs) + per-output masking arrays
3. For non-binary multi-task classification, use multiple output layers + mask arrays
4. Pass in an error array calculated externally to the network, using MultiLayerNetwork.backpropGradient(INDArray epsilon) where epsilon is (usually) dL/da, with loss function L and output activations a.

https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L2384-L2388


I'll close this for now - let me know if that is enough. Or feel free to ask me on our gitter channel.

**Comment by yylonly** [comment]
Could you provide an example for the masking arrays and manually computing the epsilon? Those are few mentioned in the documentation. Thanks
